src/macaron/slsa_analyzer/package_registry/pypi_registry.py

```diff
@@ -0,0 +1,198 @@
+        self.attestation_endpoint = attestation_endpoint or ""
+        self.request_timeout = request_timeout or 10
+        self.enabled = enabled
+        super().__init__("pypi Registry")
+
+
+class PyPIApiClient:
+    """PyPI's official and customed API client."""
+
+    def __init__(self, package: str) -> None:
+        self.headers = {"Host": "pypi.org", "Accept": "application/json"}
+        self.base_url = "https://pypi.org"
+        self.package = package
+        self.base_metadata: dict[str, Any] = self._get_base_metadata()
+
+    def _get_base_metadata(self) -> dict:
+        """Get the metadata through PyPI official API.
+
+        Returns
+        -------
+            dict | None: PyPI official metadata.
+        """
+        endpoint = os.path.join(self.base_url, "pypi", self.package, "json")
+        response: dict = send_get_http(endpoint, headers=self.headers)
+        return response
+
+    def get_releases(self) -> dict | None:
+        """Get all releases.
+
+        Returns
+        -------
+            dict | None: Version to metadata.
+        """
+        releases: dict | None = self.base_metadata.get("releases", None)
+        return releases
+
+    def get_project_links(self) -> dict | None:
+        """Get all project links.
+
+        Returns
+        -------
+            dict | None: All link's name to link.
+        """
+        project_urls: dict | None = self.base_metadata.get("info", None).get("project_urls", None)
+        return project_urls
+
+    def get_latest_version(self) -> str | None:
+        """Get latest version of the package.
+
+        Returns
+        -------
+            str | None: Latest version.
+        """
+        version: str | None = self.base_metadata.get("info", None).get("version", None)
+        return version
+
+    def get_sourcecode_url(self) -> str | None:
+        """Get the url of the source distribution.
+
+        Returns
+        -------
+            str | None: Url of the source distribution.
+        """
+        for distribution in self.base_metadata.get("urls", []):
+            if distribution.get("python_version") == "source":
+                source: str = distribution.get("url", "")
+                if source:
+                    return source
+        return None
+
+    def get_latest_release_upload_time(self) -> str | None:
+        """Get upload time of the latest release.
+
+        Returns
+        -------
+            str | None: Upload time of latest release.
+        """
+        distributions = self.base_metadata.get("urls", None)
+        return distributions[0].get("upload_time") if distributions is not None else None
+
+    def get_package_page(self) -> str | None:
+        """Implement custom API to get package main page.
+
+        Returns
+        -------
+            str | None: Package main page.
+        """
+        url = os.path.join(self.base_url, "project", self.package)
+        response = send_get_http_raw(url, headers=self.headers)
```

Comment:
@behnazh-w The `send_get_http_raw` function is performing the rate limit check using this function https://github.com/oracle/macaron/blob/31f1c87390d465d69bb24dbe5a500aafbc00e7f8/src/macaron/util.py#L115. However, the rate limit check logic here is specific for GitHub API. I think we need to do something about it :thinking: .
@Yao-Wen-Chang How does PyPI handles rate limit? Do they handle it differently to what GitHub API is doing? - https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api

Reply:
They don't have strict rules for the rate limit. 
Checkout their document https://warehouse.pypa.io/api-reference/